### DISCUSSION TITLE
Time function

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#175] [Q&A]
I take a look at the REXX TIME function, I would like to use 
      gettimeofday(&tv,&tz);
which returns me the milliseconds. 

1/ It seems gettimeofday() is not defined yet, isn't it? 

@adesutherland please provide me with some help. 

2/ my idea is to integrate it into the time instruction, distinguishing it with a second input parameter.

**Comment by adesutherland** [comment]
2/ I would prefer a mtime instruction if that is ok with you? Saves an if ...

In passing, one day we might add a ntime (nano) instruction - although this level of precision are generally being retired for security reasons (SPECTRE attacks, which work by detecting non-functional changes of the real CPU)

1/ Yes gettimeofday() is the right function for Linux and OS/X. And luckily there is a simple public domain implementation for this for Windows.

@BobBolch, if willing, will need to help with a microsecond function for VM/370 :-)

Both our instructions TIME and MTIME will get the number of seconds or microseconds since the so called epoch (00:00:00 UTC, January 1, 1970). As a deadhead I presume this was chosen in memory of the Grateful Dead playing at Fillmore East, New York, NY on January 2-3, 1970 (so one day wrong).

Anyway some maths will be needed in the REXX BIFs to get the period from year 0000.

I can make a cross-platform MTIME - but it will be Monday / Tuesday. Meantime you could just do TIME * 1000?

**Comment by Peter-Jacob** [comment]
I'll add an MTIME instruction, to use temporarily TIME() is a good idea. Therefore there is no urgency in a window version of gettimeofday(), but I'll use microseconds instead of milliseconds (nano secs would be overkill).


>> As a deadhead I presume this was chosen in memory of the Grateful Dead playing at Fillmore East
I also can't think of another explanation!

**Comment by adesutherland** [comment]
One issue I am having is that it is quite easy to get the time (seconds / milliseconds) based on GMT/UTC, but not in local time. In fact the number of seconds since 1970 for local time is kind of meaningless.

What POSIX does is have a function (localtime) to convert the number of seconds since the epoch to the year, day, hour etc. in local time.

The REXX data function was specified (and I might be wrong here) before the way to handle local time was fully developed. It expects the number of seconds since the year 0001 **in local time**. I am not sure how daylight saving time is handled short of changing the local clock. 

So I can easily get the seconds / milliseconds in UTC. I can also have an instruction to get a string (to be parsed?) with the localtime like YYYYMMDDHHMMSS, I can also get utctime like YYYYMMDDHHMMSS. Is that acceptable? The alternative is to have the REXX DATE function somehow work out local time (by reading the timezone from a TIMEZONE instruction). Hard ....

What do you guys think?

**Reply by Peter-Jacob** [reply]
I am happy with any simple portable solution. I can also live with milliseconds. I actually don't even know what precision z/OS REXX has.
If we find out we need to change it - we will change it later.

**Comment by Peter-Jacob** [comment]
Added the REXX TIME function using the C TIME function adding some fancy stuff as milliseconds. Function is committed:  96a1a841924f48ebde95706a15e0aec571f4afde 
All functions except R and E are implemented, for them, I need global variables and I couldn't get them running in REXX - @adesutherland, @rvjansen  **maybe you guys have an example for me.** 

The rvm parms are: `rvm ts_time time right left copies pos substr length`

```
say "test Time"

say "Option '' "time()
say "Option N  "time("N")
say "Option L  "time("l")
say "Option C  "time("C")
say "Option H  "time("h")
say "Option M  "time("M")
say "Option S  "time("s")

```
Result: 
```
test Time
Option '' 09:19:59
Option N  09:19:59
Option L  09:19:59.123
Option C  09:19am
Option H  9
Option M  559
Option S  33599
```

**Comment by rvjansen** [comment]
You can see globals in tests/main tests/func1 tests/func2

**Comment by rvjansen** [comment]
Did you check in the mtime source?

**Reply by Peter-Jacob** [reply]
No, I forgot 👎 ! I have done it now: 9d044a5321d824d7cd41ce79b8ced2004616fc66

**Reply by rvjansen** [reply]
great, but now:
rxc _datei
Error @ 70:42 - #31.1, "idateJ"
1 error(s) in source file

**Comment by Peter-Jacob** [comment]
I check it when I return homeVon meinem Huawei-Telefon gesendet-------- Ursprüngliche Nachricht --------Von: René Vincent Jansen ***@***.***>Datum: Mo., 1. Nov. 2021, 11:52An: adesutherland/CREXX ***@***.***>Cc: Peter-Jacob ***@***.***>, Author ***@***.***>Betreff: Re: [adesutherland/CREXX] Time function (Discussion #175)
great, but now:
rxc _datei
Error @ 70:42 - #31.1, "idateJ"
1 error(s) in source file

**Comment by Peter-Jacob** [comment]
TIME function added TIME('E') and TIME('R'), commit e326d45e8dab8d87aec121c9a2d64fd09c530f72
